Migrate kernels with Tensor? to C10 full dispatcher

wenleix opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54263 Migrate kernels with Tensor? to C10 full dispatcher**
* #54262 Add device_of overload for optional<Tensor>

Codemod commands generated by https://github.com/pytorch/pytorch/pull/54223

Signatures of the following 8 methods in LegacyTHFunctionsCUDA.h are
manually changed.

```
_thnn_multi_margin_loss_forward
_thnn_multi_margin_loss_backward
_thnn_nll_loss_forward
_thnn_nll_loss_backward
_thnn_nll_loss2d_forward
_thnn_nll_loss2d_backward
_thnn_conv2d_forward
_thnn_conv_depthwise2d_forward
```

Differential Revision: [D27164092](https://our.internmc.facebook.com/intern/diff/D27164092/)

Commit:
Migrate kernels with Tensor? to C10 full dispatcher

Codemod commands generated by https://github.com/pytorch/pytorch/pull/54223

Differential Revision: [D27164092](https://our.internmc.facebook.com/intern/diff/D27164092/)

[ghstack-poisoned]

wenleix commented on aten/src/ATen/native/miopen/Conv_miopen.cpp:
```diff
@@ -46,9 +48,11 @@ std::tuple<at::Tensor,at::Tensor,at::Tensor> miopen_convolution_backward(
+    const Tensor& input, const Tensor& weight, const c10::optional<Tensor>& bias_opt /* optional */,
     IntArrayRef padding, IntArrayRef output_padding, IntArrayRef stride, IntArrayRef dilation,
     int64_t groups, bool benchmark, bool deterministic) {
+  const Tensor& bias = c10::value_or_else(bias_opt, [] {return Tensor();});
```
Will clean up such cases (i.e. the impl just raises an error) in a separate pass.

Commit:
minor fixes on "Migrate kernels with Tensor? to C10 full dispatcher"


Codemod commands generated by https://github.com/pytorch/pytorch/pull/54223

Differential Revision: [D27164092](https://our.internmc.facebook.com/intern/diff/D27164092/)

[ghstack-poisoned]

Commit:
revert aten/src/ATen/native/xnnpack/ on "Migrate kernels with Tensor? to C10 full dispatcher"


Codemod commands generated by https://github.com/pytorch/pytorch/pull/54223

Differential Revision: [D27164092](https://our.internmc.facebook.com/intern/diff/D27164092/)

[ghstack-poisoned]

Commit:
squash & rebase on "Migrate kernels with Tensor? to C10 full dispatcher"


Codemod commands generated by https://github.com/pytorch/pytorch/pull/54223

Differential Revision: [D27164092](https://our.internmc.facebook.com/intern/diff/D27164092/)

[ghstack-poisoned]

smessmer reviewed (approved):
Looks great :)

Maybe we should also add a small comment marker into the kernels so that if we ever want to clean them up further (i.e. have the kernel handle optional directly without going through the creation of a default constructed tensor), we know how to find them.
Maybe something like
```
Tensor my_op(...) {
// See [Note: hacky_wrapper_removal]
Tensor weight = value_or_else(...)
...
}
```
and then add a [Note: hacky_wrapper_removal] somewhere in a central location.

Commit:
rebase & add note on "Migrate kernels with Tensor? to C10 full dispatcher"


Codemod commands generated by https://github.com/pytorch/pytorch/pull/54223

Signatures of the following 8 methods in LegacyTHFunctionsCUDA.h are
manually changed.

```
_thnn_multi_margin_loss_forward
_thnn_multi_margin_loss_backward
_thnn_nll_loss_forward
_thnn_nll_loss_backward
_thnn_nll_loss2d_forward
_thnn_nll_loss2d_backward
_thnn_conv2d_forward
_thnn_conv_depthwise2d_forward
```

Differential Revision: [D27164092](https://our.internmc.facebook.com/intern/diff/D27164092/)

[ghstack-poisoned]

Commit:
rebase && rework first diff based on comment on "Migrate kernels with Tensor? to C10 full dispatcher"


Codemod commands generated by https://github.com/pytorch/pytorch/pull/54223

Signatures of the following 8 methods in LegacyTHFunctionsCUDA.h are
manually changed.

```
_thnn_multi_margin_loss_forward
_thnn_multi_margin_loss_backward
_thnn_nll_loss_forward
_thnn_nll_loss_backward
_thnn_nll_loss2d_forward
_thnn_nll_loss2d_backward
_thnn_conv2d_forward
_thnn_conv_depthwise2d_forward
```

Differential Revision: [D27164092](https://our.internmc.facebook.com/intern/diff/D27164092/)

[ghstack-poisoned]

Commit:
remove adaption for unimpl stubs on "Migrate kernels with Tensor? to C10 full dispatcher"


Codemod commands generated by https://github.com/pytorch/pytorch/pull/54223

Signatures of the following 8 methods in LegacyTHFunctionsCUDA.h are
manually changed.

```
_thnn_multi_margin_loss_forward
_thnn_multi_margin_loss_backward
_thnn_nll_loss_forward
_thnn_nll_loss_backward
_thnn_nll_loss2d_forward
_thnn_nll_loss2d_backward
_thnn_conv2d_forward
_thnn_conv_depthwise2d_forward
```

Differential Revision: [D27164092](https://our.internmc.facebook.com/intern/diff/D27164092/)

[ghstack-poisoned]
